Fix arm ldrd esil ##emu ##test

SimonVareille opened the pull request:
<!-- Please read the contributing guidelines:
* https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md
In short:
* PR title must be capitalized, concise and use ##tags
* Follow the coding style, add tests and documentation if necessary
-->

**Checklist**

- [ ] Closing issues: #issue
- [x] Mark this if you consider it ready to merge
- [x] I've added tests (optional)
- [ ] I wrote some lines in the [radare2book](https://github.com/radareorg/radare2book)

**Description**

In ARM architecture, the ldrd instruction was handled as ldrb by ESIL. 
A ldrb instruction loads an unsigned byte into a register. A ldrd instruction loads a word into a register, and the following word (first word address + 4) into a second register.

The ESIL representation of ldrd was thus wrong. This PR fixes the ESIL and adds a few tests to ensure future non-regression.

`LDRB` is `LDR` with `B` type:
`LDR{type}{cond} Rt, [Rn {, #offset}] ; immediate offset`
-> Loads the byte [Rn + offset] into Rt

`LDRD` is a separate instruction:
`LDRD{cond} Rt, Rt2, [Rn {, #offset}] ; immediate offset, doubleword`
-> Loads the word [Rn + offset] into Rt and the word [Rn + offset + 4] int Rt2

trufae commented:
can  you please fix the warnings?

```
[egg 751] CC p/egg_exec.c
1177
/home/runner/work/radare2/radare2/libr/..//libr/anal/p/anal_arm_cs.c: In function ‘anop64’:
1178
/home/runner/work/radare2/radare2/libr/..//libr/anal/p/anal_arm_cs.c:3444:30: error: suggest parentheses around ‘&&’ within ‘||’ [-Werror=parentheses]
1179
 3444 |   } else if (ISPREINDEX32 () && REGBASE64 (1) == ARM64_REG_SP) {
1180
      |                              ^
1181
/home/runner/work/radare2/radare2/libr/..//libr/anal/p/anal_arm_cs.c:3447:31: error: suggest parentheses around ‘&&’ within ‘||’ [-Werror=parentheses]
1182
 3447 |   } else if (ISPOSTINDEX32 () && REGID64 (1) == ARM64_REG_SP) {
1183
      |                               ^
1184
/home/runner/work/radare2/radare2/libr/..//libr/anal/p/anal_arm_cs.c:3470:30: error: suggest parentheses around ‘&&’ within ‘||’ [-Werror=parentheses]
1185
 3470 |   } else if (ISPREINDEX32 () && REGBASE64 (1) == ARM64_REG_SP) {
1186
      |                              ^
1187
/home/runner/work/radare2/radare2/libr/..//libr/anal/p/anal_arm_cs.c:3473:31: error: suggest parentheses around ‘&&’ within ‘||’ [-Werror=parentheses]
1188
 3473 |   } else if (ISPOSTINDEX32 () && REGID64 (1) == ARM64_REG_SP) {
1189
      |                       
```

trufae commented on libr/anal/p/anal_arm_cs.c:
```diff
@@ -54,9 +55,10 @@
 #define SHIFTVALUE(x) insn->detail->arm.operands[x].shift.value
 
+#define ISWRITEBACK32() insn->detail->arm.writeback
+#define ISPREINDEX32() ((OPCOUNT () == 2) && (ISMEM (1)) && (ISWRITEBACK32 ())) || ((OPCOUNT () == 3) && (ISMEM (2)) && (ISWRITEBACK32 ()))
```
```suggestion
#define ISPREINDEX32() (((OPCOUNT () == 2) && (ISMEM (1)) && (ISWRITEBACK32 ())) || ((OPCOUNT () == 3) && (ISMEM (2)) && (ISWRITEBACK32 ())))
```

trufae commented on libr/anal/p/anal_arm_cs.c:
```diff
@@ -54,9 +55,10 @@
 
+#define ISWRITEBACK32() insn->detail->arm.writeback
+#define ISPREINDEX32() ((OPCOUNT () == 2) && (ISMEM (1)) && (ISWRITEBACK32 ())) || ((OPCOUNT () == 3) && (ISMEM (2)) && (ISWRITEBACK32 ()))
+#define ISPOSTINDEX32() ((OPCOUNT () == 3) && (ISIMM (2) || ISREG (2)) && (ISWRITEBACK32 ())) || ((OPCOUNT () == 4) && (ISIMM (3) || ISREG (3)) && (ISWRITEBACK32 ()))
```
```suggestion
#define ISPOSTINDEX32() (((OPCOUNT () == 3) && (ISIMM (2) || ISREG (2)) && (ISWRITEBACK32 ())) || ((OPCOUNT () == 4) && (ISIMM (3) || ISREG (3)) && (ISWRITEBACK32 ())))
```

trufae commented:
1 failing test! Can you fix it? And verify the new output is correct?

![9BFC0BE4-AD4F-465E-BB21-D7D95EC7D60E](https://user-images.githubusercontent.com/6431515/120230850-77eb1780-c250-11eb-9f3f-dbe2dfec4940.jpeg)

trufae commented:
![Uploading Screenshot 2021-05-31 at 21.04.35.png…]()

trufae commented:
![Screenshot 2021-05-31 at 21 04 35](https://user-images.githubusercontent.com/6431515/120232430-db2a7900-c253-11eb-9de0-f268d0dddc21.png)

SimonVareille commented:
Well... If I checkout to a609fb57d64f37107ad9ee628d44c23f34c8e8da (just before my first commit) the test fails...

On master the test fails too...

radare commented:
master is green, and this test passes fine in local for me

radare commented:
i have updated my previous comment with some details on it

radare commented:
Here's the binary i cooked for testing this 
[hello-ios-arm64.zip](https://github.com/radareorg/radare2/files/6572047/hello-ios-arm64.zip)

SimonVareille commented:
Ok, it should be fixed now.

The problem was that ARM64 instructions were using ISPREINDEX32 and ISPOSTINDEX32 macros, which I modified to be properly used by ARM32 instructions. There were ISPREINDEX64 and ISPOSTINDEX64 macros before, but they where not complete.

The previous implementation of ISPREINDEX32 and ISPOSTINDEX32 was coherent with a use of 64 bits instructions. Now, I updated ISPREINDEX64 and ISPOSTINDEX64 to be used in all required cases.

trufae commented:
Let's wait 15m for linux-test to finish but looks good! nice catch

trufae commented:
The var/arg analysis issue should be addressed in a separate PR and i think it's unrelated to how esil is generated.

trufae commented on libr/anal/p/anal_arm_cs.c:
```diff
@@ -1690,13 +1692,13 @@ static int analop64_esil(RAnal *a, RAnalOp *op, ut64 addr, const ut8 *buf, int l
 				// I assume the DUPs here previously were to handle preindexing
 				// but it was never finished?
-				if (ISPREINDEX32()) {
+				if (ISPREINDEX64()) {
```
```suggestion
				if (ISPREINDEX64 ()) {
```

trufae commented on libr/anal/p/anal_arm_cs.c:
```diff
@@ -1690,13 +1692,13 @@ static int analop64_esil(RAnal *a, RAnalOp *op, ut64 addr, const ut8 *buf, int l
 				r_strbuf_appendf (&op->esil, ",[%d],%s,=", size, REG64 (0));
 
-				if (ISPOSTINDEX32()) {
+				if (ISPOSTINDEX64()) {
```
```suggestion
				if (ISPOSTINDEX64 ()) {
```

trufae commented on libr/anal/p/anal_arm_cs.c:
```diff
@@ -1783,13 +1785,13 @@ static int analop64_esil(RAnal *a, RAnalOp *op, ut64 addr, const ut8 *buf, int l
 				// I assume the DUPs here previously were to handle preindexing
 				// but it was never finished?
-				if (ISPREINDEX32()) {
+				if (ISPREINDEX64()) {
```
```suggestion
				if (ISPREINDEX64 ()) {
```

trufae commented on libr/anal/p/anal_arm_cs.c:
```diff
@@ -1783,13 +1785,13 @@ static int analop64_esil(RAnal *a, RAnalOp *op, ut64 addr, const ut8 *buf, int l
 				r_strbuf_appendf (&op->esil, ",[%d],~,%s,=", size, REG64 (0));
 				
-				if (ISPOSTINDEX32()) {
+				if (ISPOSTINDEX64()) {
```
```suggestion
				if (ISPOSTINDEX64 ()) {
```

trufae commented on libr/anal/p/anal_arm_cs.c:
```diff
@@ -1911,13 +1913,13 @@ static int analop64_esil(RAnal *a, RAnalOp *op, ut64 addr, const ut8 *buf, int l
 				// I assume the DUPs here previously were to handle preindexing
 				// but it was never finished?
-				if (ISPREINDEX32()) {
+				if (ISPREINDEX64()) {
```
```suggestion
				if (ISPREINDEX64 ()) {
```

trufae commented on libr/anal/p/anal_arm_cs.c:
```diff
@@ -1911,13 +1913,13 @@ static int analop64_esil(RAnal *a, RAnalOp *op, ut64 addr, const ut8 *buf, int l
 				r_strbuf_appendf (&op->esil, ",=[%d]", size);
 
-				if (ISPOSTINDEX32()) {
+				if (ISPOSTINDEX64()) {
```
```suggestion
				if (ISPOSTINDEX64 ()) {
```

trufae commented on libr/anal/p/anal_arm_cs.c:
```diff
@@ -1911,13 +1913,13 @@ static int analop64_esil(RAnal *a, RAnalOp *op, ut64 addr, const ut8 *buf, int l
 				r_strbuf_appendf (&op->esil, ",=[%d]", size);
 
-				if (ISPOSTINDEX32()) {
+				if (ISPOSTINDEX64()) {
```
```suggestion
				if (ISPOSTINDEX64 ()) {
```

trufae commented:
Merged! Good contrib! 

i filled an issue in https://github.com/radareorg/radare2/issues/18788 to track the var/arg analysis issue
